### PR TITLE
Improve validation on dictionary parameters in SetSample

### LINKED ISSUE #30621: Can't run SetSample from algorithm window without specifying an Environment
It's not possible to leave the Environment parameter blank when running the SetSample algorithm from the algorithm window even though the parameter is non-mandatory.

Mantid requires a valid json string for the parameter. If you try to work around this by supplying a pair of empty curly braces that also doesn't work because you get a validation error saying the Name parameter is required.

The validation should be relaxed so you can just leave this field empty.

This works OK when running the algorithm from a python script

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

When SetSample was run from the Algorithm dialog in the workbench UI it wasn't possible to leave some of the dictionary parameters blank. This is a valid use case for some combinations of parameters so this PR allows that.

Also when clearing a parameter that previously had a value the "blanking out" didn't take effect straight away. You had to reopen the algorithm dialog and try to run the algorithm a second time to see validation errors relating to empty fields

Made one other minor change to one of the SetSampleTest unit tests - I noticed TS_ASSERT_THROWS doesn't seem to check the exception type when running on a release build in Visual Studio. So the test was failing in debug. My change sorts that out

**To test:**

Run the SetSample line of the following script via the Algorithm dialog in Workbench (replacing the single quotes with double quotes - the algorithm dialog seems to require this for some reason):

```
ws = CreateSampleWorkspace()
SetSample(ws, Geometry={'Shape': 'HollowCylinder', 'Height': 4.0, 'InnerRadius': 0.8, 'OuterRadius': 1.0, 'Center': [0.,0.,0.]},
          Material={'ChemicalFormula': '(Li7)2-C-H4-N-Cl6','NumberDensity': 0.1})
```

Leaving the environment parameter blank wasn't possible before

Fixes #30621.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by StephenSmith25 — COMMENTED** [review]
Looks good, fixes the issue - Just two minor comments.

File: Framework/DataHandling/inc/MantidDataHandling/SetSample.h [hunk 1]
```diff
@@ -79,6 +79,7 @@ class MANTID_DATAHANDLING_DLL SetSample final : public API::Algorithm {
                    const Kernel::PropertyManager &materialArgs);
   Kernel::PropertyManager materialSettingsEnsureLegacyCompatibility(
       const Kernel::PropertyManager &materialArgs);
+  bool isDictionaryPopulated(const Kernel::PropertyManager_const_sptr dict);
```
[the remark this answers is not in the document]
Comment (DannyHindson) [reply Framework/DataHandling/inc/MantidDataHandling/SetSample.h:82]: Yes I'll change it to a const method. 
Re the shared pointer - won't the use count get decremented at the end of the call to the method? Or perhaps you were more concerned about the performance from unnecessarily incrementing\decrementing the count...
I just looked at our coding standards and it suggests not passing shared_ptr unless transferring ownership. So perhaps in this case I should just pass a raw pointer? I'd like to include the check on whether the pointer is null or not in the method so don't really want to pass the object itself by reference.

[the remark this answers is not in the document]
Comment (DannyHindson) [reply Framework/DataHandling/inc/MantidDataHandling/SetSample.h:82]: Actually - just tried changing the method to take a raw pointer and that will involve me putting .get() on the end of all the small pointers in order to call the method so not so keen on that. I'll do what you suggested 😃 and pass the smart pointer by reference which seems like it should avoid any performance pitfalls from passing the smart pointer.

**Review by StephenSmith25 — APPROVED** [review]
Great, looks good! 👍

**Comment by DannyHindson** [comment]
> Great, looks good! 👍

Thanks. For some reason the windows build took a whopping 5 hours 36 mins but it got there in the end!

**Comment by StephenSmith25** [comment]
Yeah, I was wondering about that - couldn't see anything wrong with the build, so no idea why it took that long! Maybe something to keep an eye on!

**Pull request merged**